File: src/ctci6/utils/BinaryTreeNode.java
package ctci6.utils;

/**
 * Created by hao on 10/3/16.
 */
public class BinaryTreeNode<Value> {

    public Value val;
    public BinaryTreeNode left;
    public BinaryTreeNode right;
    public BinaryTreeNode parent;

    public BinaryTreeNode(Value val) {
        this.val = val;
    }

    public BinaryTreeNode sibling() {
        if (parent == null) return null;
        BinaryTreeNode sibling = parent.left == this ? parent.right : parent.left;
        return sibling;
    }

    public int size() {
        return size(this);
    }

    private int size(BinaryTreeNode node) {
        if (node == null) return 0;
        return size(node.left) + size(node.right) + 1;
    }
}

File: ctci6/utils/BinaryTreeNode.js
var Queue = require('./Queue');

class BinaryTreeNode {
    constructor (val) {
        this.val = val;
        this.left = null;
        this.right = null;
        this.d = 0; // depth
        this.h = 0; // height
    }

    height() {
        var max = 0;
        var q = new Queue();
        q.enqueue(this);
        while (!q.isEmpty()) {
            var node = q.dequeue();
            if (node.d > max) max = node.d;
            var left = node.left;
            var right = node.right;
            if (left) {
                left.d = node.d + 1;
                q.enqueue(left);
            }
            if (right) {
                right.d = node.d + 1;
                q.enqueue(right)
            }
        }
        return max;
    }
}

module.exports = BinaryTreeNode;

function test() {
    var node = new BinaryTreeNode(1);
    var a = new BinaryTreeNode(2);
    var b = new BinaryTreeNode(3);
    node.left = a;
    node.right = b;
    console.log(node.height());
}

if (require.main === module) test();